### ISSUE TITLE
Fatal Error on ggt sync

### ISSUE THREAD

**Issue by sb18281** [gadget-inc/ggt#18]
Not 100% sure if there was a specific action that caused it or if it was time-based. 

Here's the error info
<details>
<summary>Error Message</summary>

```
<--- Last few GCs --->

[199432:0x4c30030]   786320 ms: Mark-sweep 4041.9 (4138.1) -> 4029.9 (4142.1) MB, 3379.8 / 8.0 ms  (average mu = 0.218, current mu = 0.006) allocation failure scavenge might not succeed
[199432:0x4c30030]   790423 ms: Mark-sweep 4045.8 (4142.1) -> 4033.6 (4145.6) MB, 4079.3 / 7.2 ms  (average mu = 0.114, current mu = 0.006) allocation failure scavenge might not succeed


<--- JS stacktrace --->

FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory
 1: 0xb17ec0 node::Abort() [node]
 2: 0xa341f4 node::FatalError(char const*, char const*) [node]
 3: 0xcfe71e v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [node]
 4: 0xcfea97 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [node]
 5: 0xee8d35  [node]
 6: 0xef7ab1 v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [node]
 7: 0xefad0c v8::internal::Heap::AllocateRawWithRetryOrFailSlowPath(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [node]
 8: 0xec72bb v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [node]
 9: 0x123052b v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [node]
10: 0x16147d9  [node]
[1]    199432 abort (core dumped)  ggt sync --app contentful-shopify-migrator .
```

</details>

<details>
<summary>Prior console output</summary>

```
ggt sync --app contentful-shopify-migrator .
Ready
Received
← .gadget/client/dist-cjs/Client.d.ts
← .gadget/client/dist-cjs/Client.js
← .gadget/client/dist-cjs/Client.js.map
← .gadget/client/dist-cjs/iife-export.d.ts
← .gadget/client/dist-cjs/iife-export.js
← .gadget/client/dist-cjs/iife-export.js.map
← .gadget/client/dist-cjs/index.d.ts
← .gadget/client/dist-cjs/index.js
← .gadget/client/dist-cjs/index.js.map
← .gadget/client/dist-cjs/models/ContentfulArticle.d.ts
… 245 more
Received
← .gadget/client/dist-cjs/Client.d.ts
← .gadget/client/dist-cjs/iife-export.d.ts
← .gadget/client/dist-cjs/index.d.ts
← .gadget/client/dist-cjs/models/ContentfulArticle.d.ts
← .gadget/client/dist-cjs/models/ContentfulArticleProcessQueue.d.ts
← .gadget/client/dist-cjs/models/ContentfulImageContent.d.ts
← .gadget/client/dist-cjs/models/ContentfulTextContent.d.ts
← .gadget/client/dist-cjs/models/CurrentSession.d.ts
← .gadget/client/dist-cjs/models/RelatedContentfulArticles.d.ts
← .gadget/client/dist-cjs/models/RelatedProducts.d.ts
… 26 more
Received
← .gadget/client/dist-cjs/Client.js
← .gadget/client/dist-cjs/Client.js.map
← .gadget/client/dist-cjs/index.js.map
← .gadget/client/dist-cjs/types.js.map
← .gadget/client/dist-esm/Client.js
← .gadget/client/dist-esm/Client.js.map
← .gadget/client/dist-esm/index.js.map
← .gadget/client/dist-esm/types.js.map
← .gadget/client/package.json
← .gadget/client/src/Client.ts
… 2 more
Received
← .gadget/server/dist/GlobalActionContexts.d.ts
← .gadget/server/dist/GlobalActionContexts.js
← .gadget/server/dist/GlobalActionContexts.js.map
← .gadget/server/dist/index.d.ts
← .gadget/server/package.json
← .gadget/server/src/GlobalActionContexts.ts
← .gadget/server/src/index.ts
Received
← .gadget/client/dist-cjs/Client.d.ts
← .gadget/client/dist-cjs/iife-export.d.ts
← .gadget/client/dist-cjs/index.d.ts
← .gadget/client/dist-cjs/models/ContentfulArticle.d.ts
← .gadget/client/dist-cjs/models/ContentfulArticleProcessQueue.d.ts
← .gadget/client/dist-cjs/models/ContentfulImageContent.d.ts
← .gadget/client/dist-cjs/models/ContentfulTextContent.d.ts
← .gadget/client/dist-cjs/models/CurrentSession.d.ts
← .gadget/client/dist-cjs/models/RelatedContentfulArticles.d.ts
← .gadget/client/dist-cjs/models/RelatedProducts.d.ts
… 12 more
Received
← global/processQueue/ProcessQueue.ts
Received
← .gadget/client/dist-cjs/Client.d.ts
← .gadget/client/dist-cjs/iife-export.d.ts
← .gadget/client/dist-cjs/index.d.ts
← .gadget/client/dist-cjs/models/ContentfulArticle.d.ts
← .gadget/client/dist-cjs/models/ContentfulArticleProcessQueue.d.ts
← .gadget/client/dist-cjs/models/ContentfulImageContent.d.ts
← .gadget/client/dist-cjs/models/ContentfulTextContent.d.ts
← .gadget/client/dist-cjs/models/CurrentSession.d.ts
← .gadget/client/dist-cjs/models/RelatedContentfulArticles.d.ts
← .gadget/client/dist-cjs/models/RelatedProducts.d.ts
… 16 more
Received
← .gadget/server/dist/index.d.ts
← .gadget/server/package.json
← .gadget/server/src/index.ts
Received
← .gadget/client/dist-cjs/Client.d.ts
← .gadget/client/dist-cjs/iife-export.d.ts
← .gadget/client/dist-cjs/index.d.ts
← .gadget/client/dist-cjs/models/ContentfulArticle.d.ts
← .gadget/client/dist-cjs/models/ContentfulArticleProcessQueue.d.ts
← .gadget/client/dist-cjs/models/ContentfulImageContent.d.ts
← .gadget/client/dist-cjs/models/ContentfulTextContent.d.ts
← .gadget/client/dist-cjs/models/CurrentSession.d.ts
← .gadget/client/dist-cjs/models/RelatedContentfulArticles.d.ts
← .gadget/client/dist-cjs/models/RelatedProducts.d.ts
… 12 more
Sent
→ yarn.lock
→ package.json
Sent
→ package.json
Sent
→ package.json

```
</details>

Device Info 
- Ubuntu 20.04.4 LTS
- Node v16.3.0
- Yarn 1.22.18

I tried to restart the sync command and received a `GGT_CLI_INVALID_SYNC_FILE` error, missing `.gadget/sync.json`

**Comment by sb18281** [comment]
Seems to have occured again on a different machine (similar specs), but only the first run. On both machines I was able to re-sync usign force after the first run

**Comment by angelini** [comment]
Hey @sb18281 thanks for the useful bug report.

The error itself is the Node VM running out of memory, by default it is given 4GB. The sync tool should have no reason for using that much memory, so clearly we're leaking something. We had already begun to investigate higher than expected memory usage over time and will post any updates here.

> I tried to restart the sync command and received a GGT_CLI_INVALID_SYNC_FILE error, missing .gadget/sync.json

This is an unfortunate consequence of the way the VM crashes when it runs out of memory. We write out the state of the sync when the program exits, successfully or by exception handler. However in the case of a memory leak the VM crashes and we can't write out the `sync` file.

/cc @scott-rc

**Comment by scott-rc** [comment]
Hey @sb18281, sorry for the long delay! I finally figured out and released a fix for this issue today in version `v0.1.4`.

You can install the latest version of `ggt` via:

```sh
npm install -g @gadgetinc/ggt
```

To verify the correct version is installed:

```sh-session
$ ggt --version
@gadgetinc/ggt/0.1.4 linux-arm64 node-v16.17.1
```

---

The issue was fixed via #21.

We use [Chokidar](https://github.com/paulmillr/chokidar) to watch your local filesystem. Chokidar consumes more memory as the amount of files it watches increases. Chokidar has an option to ignore certain paths so that it doesn't watch unnecessary files, however, we were passing the paths incorrectly so Chokidar was watching _everything_ -- including `node_modules` 😱. Since `node_modules` [is huge](https://preview.redd.it/tfugj4n3l6ez.png?auto=webp&s=b8163176d8482d5e78ac631e16b7973a52e3b188) and Chokidar was watching all the files inside of it, that is what was causing the `JavaScript heap out of memory` issue.

If the issue persists, even after upgrading to the latest version of `ggt`, feel free to re-open this issue.

Cheers!
